ivadomed/ivadomed
# Add Multi-Session/Contrasts Support to Ivadomed

## Checklist

#### GitHub




#### PR contents


## Description


This PR is the first attempt address multi-session and contrasts compatibility issue to the existing `ivadomed` process by significantly overhaul the 

### Preamble
   * This PR fundamentally change the nature of how indexing of the derivatives files work. Those changes are concentrated in the `ivadomed/loader/bids_dataframe.py` with its significantly revised multi stage filtering process for bids dataframes. Before examining the detailed changes, **I recommend to review the new unit tests at `test_loader_multi`** to see if those cases and the expected output (which you can get form the data_test_multi repo linked here: 
   * https://github.com/MotionCorrect/data_multi-sessions-contrasts/archive/refs/tags/v2022-01-06.zip
   * Beside that, this PR also add a few minor things here and there that was used to help develop it
      * Add [stepping debugging unit test to automate_training to enable step wise debugging](https://github.com/ivadomed/ivadomed/pull/1044/files#diff-8b2d531f36d5b0a30be8abc33bf837becdebede8930610c0d325bc9b40c39777R38) by calling its core function directly instead of via CLI/API to enable easier debugging during development
      * Adopt [pytest-cases](https://github.com/ivadomed/ivadomed/pull/1044/files#diff-c9c796259f3852b51b531b79cbf07820145088d4f108fb006ab44b15f0b15934R9) for easier test case scenario management process. Another PR is being prepared by @konantian to revamp it for ALL unit/functional tests to streamline the adopt the pytest-cases. https://github.com/ivadomed/ivadomed/pull/1071
      * Adopt standardized input tracing to allow printing of loader parameter during development process WHILE showing nothing to end user without any code change by leveraging [Loguru's environmental variable](https://loguru.readthedocs.io/en/stable/api/logger.html#env) 

### Update to Download Script
  * Note that for this PR to be reviewed while minimizing interruptions to the data testing repos (both regular one at https://github.com/MotionCorrect/data-testing/archive/refs/tags/r20220116.zip and https://github.com/MotionCorrect/data_multi-sessions-contrasts/archive/refs/tags/v2022-01-06.zip release versions are used instead. 
  * The download scripts have been temporarily updated to point to these two sources to enable download/testing. When PR is approved here before merging, those two data repos will be created and the download script URL will be updated accordingly. 

### How to test

   * Clone down the source branch
   * **Reinstall requirement_dev.txt**
   * Make sure all local unit test are passing
   * Review the newly added unit tests in the `testing/unit_tests/test_loader_multi.py` and see if those `cases` are providing expected output. 
   * **Discuss about what you like and what you don't like**. ESPECIALLY CHECK IF THE TEST CASES behave not as you expected (i.e. no need to dig into the lower level code at implementation level if at behavioral level if you don't agree with it)
   * Bring any cause for concern to the `ivadomed` meeting

### Major Components to Be Reviewed in Detail Once Indexing Behaviour is Agreed:
 
   * `ivadomed/loader/bids_dataframe.py`
      * Significantly refactored and updated. 
      * `create_bids_dataframe` has been significantly changed  such that there are now THREE passes over the data set to refine the data frame: 
          *  first: `self.first_inclusive_pass_data_frame_creation()`
          * second: `self.second_exclusive_pass_dataframe_creation_removing_subjects_without_derivatives(self.df)`
          * third: `self.third_exclusive_pass_df_creation_check_modalities_sessions_combinations(self.df)`

   * `ivadomed/loader/bids_dataset.py`
      * Significantly refactored and updated. 
      * Refactored and created: `validate_derivative_path_to_update_target_filename` function
      * Refactored and `create_filename_pair()` function

   * `ivadomed/main.py`
      * Session specific handling added from ~285 onwards

   * `ivadomed/utils.py` 
      * adding similarity score metrics to help conduct string comparisons used to distinguish multiple similar ground truth files if they exist. 
 
### Testing Related Major Changes

   * `ivadomed/scripts/download_data.py`
      * Added data_testing and added data_multi_testing
      * To be updated in follow up PR to use official ivadomed data-testing and data-multi-session-contrasts testing releases once main master PR has been merged. 

  * `requirements_dev.txt`
      * adding pytest-cases used in data_multi_testing related unit testing processes. 

   * `testing/common_testing_util.py`
      *  Add a common function to compare dataframe comparison frameworks
      * Add a generic method for generating common/default config FROM the `config.json` file via dictionary overloading to update the default template

   * `testing/functional_tests/test_automate_training.py`
      * Add an ability to step into the high level organized script and its underlying function calls of automated_training, enabling its step wise debugging functionalities (normal unit tests call a script via CLI and cannot be stepped into. 

   * `testing/unit_tests/test_loader_multi.py`
      * Add `pytest-case` based case-specific driven testing for multi-contrast/multi-session unit testing for several major foreseeable scenarios with the varying inclusion of ground truth, sessions and contrast. 

   * `testing/unit_tests/cases_multi_ground_truth.py`
      * Defining cases for single, multiple, null, undefined **ground truth** specification in loader parameters. 

   * `testing/unit_tests/cases_multi_sessions.py`
      * Defining cases for single, multiple, null, undefined **sessions** specification in loader parameters. 

   * `testing/unit_tests/cases_multi_contrasts.py`
      * Defining cases for single, multiple, null, undefined **contrasts** specification in loader parameters.  

   * `testing/unit_tests/t_utils.py`
      * Add pytest fixtures to automatically download multi-session/contrasts related data. 

## Linked issues


 * Relevant issues: 
    * #788
    * #808
    * #1009 
    * and other issues listed under Project 5: https://github.com/ivadomed/ivadomed/projects/5
